wit/caching.wit

```diff
@@ -0,0 +1,83 @@
+    type future-get-result = u32
+    drop-future-get-result: func(f: future-get-result)
+    future-get-result-get: func(f: future-get-result) -> option<result<option<incoming-value>, error>>
+    listen-to-future-get-result: func(f: future-get-result) -> pollable
+
+    // The `exists` operation returns whether a value was previously `set` for
+    // the given key within the TTL.
+    exists: func(k: key) -> future-exists-result
+
+    // This block defines a special resource type used by `exists` to emulate
+    // `future<result<bool,error>>`.
+    type future-exists-result = u32
+    drop-future-exists-result: func(f: future-exists-result)
+    future-exists-result-get: func(f: future-exists-result) -> option<result<bool, error>>
+    listen-to-future-exists-result: func(f: future-exists-result) -> pollable
+
+    // The `set` operation sets the given value for the given key for the given
+    // time-to-live (TTL) duration, if supplied, specified in milliseconds. If
+    // the key is already present in the cache, the value is updated in-place.
+    // In the common case of computing and caching a value if the given key is
+    // not already in the cache, consider using `get-or-set` (below) intead of
+    // separate `get` and `set` operations.
+    set: func(k: key, v: outgoing-value, TTL-ms: option<u32>) -> future-set-result
+
+    // This block defines a special resource type used by `set` to emulate
+    // `future<result<_,error>>`.
+    type future-set-result = u32
+    drop-future-set-result: func(f: future-set-result)
+    future-set-result-get: func(f: future-set-result) -> option<result<_, error>>
+    listen-to-future-set-result: func(f: future-set-result) -> pollable
+
+    // The `get-or-set` operation asynchronously returns one of two cases
+    // enumerated by `get-or-set-entry`: in the `occupied` case, the given key
+    // already has a value present in the cache; in the `vacant` case, there
+    // was no value and the caller should write a value into the returned
+    // `outgoing-value` resource. This operation allows multiple concurrent
+    // `get-or-set` invocations to rendezvous such that only one invocation
+    // receives the `vacant` result while all other invocations wait until the
+    // `outgoing-value` is written before receiving an `occupied` result.
+    // Implementations are not required to implement this rendezvous or to
+    // rendezvous in all possible cases. If a caller receives a `vacant` result
+    // and drops the `outgoing-value` before completely writing the value, the
+    // `set` is cancelled and another `get-or-set` operation must be unblocked
+    // by returning `vacant`. Implementations may impose a timeout on writing
+    // an `outgoing-value` that also automatically cancels the `set`.
+    variant get-or-set-entry {
+      occupied(incoming-value),
+      vacant(outgoing-value)
+    }
+    get-or-set: func(k: key, TTL-ms: option<u32>) -> future-get-or-set-result
```

Comment:
It would be ideal if the TTL could be provided after the point where the client knows it has a `vacant` in hand. Often the TTL will be determined based on a resource that is fetched in order to fill the `vacant` slot (such as a `cache-control: max-age` header). Maybe `vacant` could be a callback along the lines of `func(TTL-ms: option<u32>) -> outgoing-value`?

Reply:
Ah, very useful feedback, thanks; I'll take a stab that.

Reply:
[Fixed](https://github.com/WebAssembly/wasi-keyvalue/pull/16/commits/64210230e41e73520ff8773ef67b266c61450250), PTAL

Reply:
Is _this_ `TTL-ms` param still needed?

Reply:
Hah, right you are; [fixed](https://github.com/WebAssembly/wasi-keyvalue/pull/16/commits/b2e9bab71fa7e9f1fbe88c9a57d82988321f4744)

Reply:
Looks great, thank you!